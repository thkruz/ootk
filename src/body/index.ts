export { Celestial } from './Celestial';
export { Earth } from './Earth';
export { Moon } from './Moon';
export { Sun } from './Sun';
