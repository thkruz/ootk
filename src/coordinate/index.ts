export * from './Hill.js';
export * from './ClassicalElements.js';
export * from './EquinoctialElements.js';
export * from './FormatTle.js';
export * from './Geodetic.js';
export * from './ITRF.js';
export * from './J2000.js';
export * from './RelativeState.js';
export * from './RIC.js';
export * from './StateVector.js';
export * from './TEME.js';
export * from './Tle.js';
