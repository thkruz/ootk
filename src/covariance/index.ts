export * from './CovarianceSample.js';
export * from './StateCovariance.js';
