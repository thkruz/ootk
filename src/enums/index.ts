export { CatalogSource } from './CatalogSource.js';
export { CommLink } from './CommLink.js';
export { OrbitRegime } from './OrbitRegime.js';
export { Sgp4OpsMode } from './Sgp4OpsMode.js';
export { AngularDiameterMethod } from './AngularDiameterMethod.js';
export { AngularDistanceMethod } from './AngularDistanceMethod.js';
export { PassType } from './PassType.js';
