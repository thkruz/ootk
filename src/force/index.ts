export { EarthGravity } from './EarthGravity.js';
export { Force } from './Force.js';
export { ForceModel } from './ForceModel.js';
export { Gravity } from './Gravity.js';
export { SolarRadiationPressure } from './SolarRadiationPressure.js';
export { ThirdBodyGravity } from './ThirdBodyGravity.js';
export { Thrust } from './Thrust.js';

