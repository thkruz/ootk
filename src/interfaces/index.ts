export type { GroundPositionParams } from './GroundPositionParams.js';
export type { OptionsParams } from './OptionsParams.js';
export type { TimeVariables } from './TimeVariables.js';
export type { StarObjectParams } from './StarObjectParams.js';
export type { ClassicalElementsParams } from './ClassicalElementsParams.js';
export type { EquinoctialElementsParams } from './EquinoctialElementsParams.js';
export type { ScenarioParams } from '../objects/Scenario.js';
export type { BaseObjectParams } from '../objects/base/BaseObject.js';
export type { GroundObjectParams } from '../objects/base/GroundObject.js';
export type { FacilityParams } from '../objects/base/Facility.js';
export type { SensorParams } from '../objects/attachable/SensorAdv.js';
export type { RectangularSensorParams } from '../objects/attachable/RectangularSensor.js';
