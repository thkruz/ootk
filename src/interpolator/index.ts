export { ChebyshevCoefficients } from './ChebyshevCoefficients.js';
export { ChebyshevInterpolator } from './ChebyshevInterpolator.js';
export { CubicSpline } from './CubicSpline.js';
export { CubicSplineInterpolator } from './CubicSplineInterpolator.js';
export { Interpolator } from './Interpolator.js';
export { LagrangeInterpolator } from './LagrangeInterpolator.js';
export { StateInterpolator } from './StateInterpolator.js';
export { VerletBlendInterpolator } from './VerletBlendInterpolator.js';

