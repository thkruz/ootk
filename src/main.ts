/**
 * @author @thkruz Theodore Kruczek
 * @description Orbital Object ToolKit (ootk) is a collection of tools for working
 * with satellites and other orbital objects.
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Many of the classes are based off of the work of @david-rc-dayton and his
 * Pious Squid library (https://github.com/david-rc-dayton/pious_squid).
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './time/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
export * from './operations/operations.js';
export { BaseObject, Sensor } from './objects/index.js';
export { Earth } from './body/index.js';
export * from './coordinate/index.js';
export * from './observation/index.js';
export * from './data/DataHandler.js';
export * from './sgp4/index.js';
export * from './objects/index.js';
export * from './interpolator/index.js';
export * from './force/index.js';
export * from './maneuver/index.js';
export * from './optimize/index.js';
export * from './orbit_determination/index.js';
export * from './body/index.js';
export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './coordinate/index.js';
export * from './operations/index.js';
export * from './force/index.js';
export * from './propagator/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
