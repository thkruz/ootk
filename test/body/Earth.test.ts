import { Earth, EpochUTC, Kilometers, Vector3D } from '../../src/main';

describe('Earth', () => {
  let exampleDate = new Date(1705109326817);

  beforeEach(() => {
    const mockExampleDate = new Date(1705109326817);

    exampleDate = new Date(mockExampleDate.getTime());
  });

  // can calculate mean motion from semimajor axis
  it('should calculate mean motion when given a semimajor axis', () => {
    const semimajorAxis = 7000 as Kilometers;
    const meanMotion = Earth.smaToMeanMotion(semimajorAxis);

    expect(meanMotion).toMatchSnapshot();
  });

  // can calculate precession angles from epoch
  it('should calculate precession angles when given an epoch', () => {
    const epoch = EpochUTC.fromDateTime(exampleDate);
    const precessionAngles = Earth.precession(epoch);

    expect(precessionAngles.zeta).toMatchSnapshot();
    expect(precessionAngles.theta).toMatchSnapshot();
    expect(precessionAngles.zed).toMatchSnapshot();
  });

  // can calculate nutation angles from epoch
  it('should calculate nutation angles when given an epoch', () => {
    const epoch = EpochUTC.fromDateTime(exampleDate);
    const nutationAngles = Earth.nutation(epoch);

    expect(nutationAngles.dPsi).toMatchSnapshot();
    expect(nutationAngles.dEps).toMatchSnapshot();
    expect(nutationAngles.mEps).toMatchSnapshot();
    expect(nutationAngles.eps).toMatchSnapshot();
    expect(nutationAngles.eqEq).toMatchSnapshot();
    expect(nutationAngles.gast).toMatchSnapshot();
  });

  // can handle semimajor axis of 0 when calculating mean motion
  it('should return NaN when given a semimajor axis of 0', () => {
    const semimajorAxis = 0 as Kilometers;
    const meanMotion = Earth.smaToMeanMotion(semimajorAxis);

    expect(meanMotion).toBe(Infinity);
  });

  // can handle negative semimajor axis when calculating mean motion
  it('should return NaN when given a negative semimajor axis', () => {
    const semimajorAxis = -7000 as Kilometers;
    const meanMotion = Earth.smaToMeanMotion(semimajorAxis);

    expect(meanMotion).toBeNaN();
  });

  // can calculate drift rate from semimajor axis
  it('should calculate drift rate when given a semimajor axis', () => {
    const semimajorAxis = 7000; // km
    const driftRate = Earth.smaToDrift(semimajorAxis);

    expect(driftRate).toMatchSnapshot();
  });

  // can calculate semimajor axis from drift rate
  it('should calculate semimajor axis from drift rate', () => {
    const driftRate = 0.1; // degrees per day
    const semimajorAxis = Earth.driftDegreesToSma(driftRate);

    expect(semimajorAxis).toMatchSnapshot();
  });

  // can calculate Earth's diameter from satellite position
  it('should calculate Earth\'s diameter from satellite position', () => {
    const satPos = new Vector3D(1000, 1000, 1000);
    const diameter = Earth.diameter(satPos);

    expect(diameter).toMatchSnapshot();
  });

  // can handle epoch before 1900 when calculating precession angles
  it('should handle epoch before 1900 when calculating precession angles', () => {
    const epoch = EpochUTC.fromDateTime(exampleDate);
    const precessionAngles = Earth.precession(epoch);

    expect(precessionAngles.zeta).toMatchSnapshot();
  });

  // can handle epoch after 2150 when calculating precession angles
  it('should handle epoch after 2150 when calculating precession angles', () => {
    const epoch = EpochUTC.fromDateTime(exampleDate);
    const precessionAngles = Earth.precession(epoch);

    expect(precessionAngles.zeta).not.toBeNaN();
    expect(precessionAngles.zeta).toMatchSnapshot();
  });

  // can handle epoch before 1900 when calculating nutation angles
  it('should handle epoch before 1900 when calculating nutation angles', () => {
    const epoch = EpochUTC.fromDateTime(exampleDate);
    const nutationAngles = Earth.nutation(epoch);

    expect(nutationAngles.dPsi).not.toBeNaN();
    expect(nutationAngles.dPsi).toMatchSnapshot();
  });

  // can handle epoch after 2150 when calculating nutation angles
  it('should handle epoch after 2150 when calculating nutation angles', () => {
    const epoch = EpochUTC.fromDateTime(exampleDate);
    const nutationAngles = Earth.nutation(epoch);

    expect(nutationAngles.dPsi).not.toBeNaN();
    expect(nutationAngles.dPsi).toMatchSnapshot();
  });

  // can handle negative drift rate when calculating semimajor axis
  it('should handle negative drift rate when calculating semimajor axis', () => {
    const driftRate = -0.1; // degrees per day
    const semimajorAxis = Earth.driftDegreesToSma(driftRate);

    expect(semimajorAxis).not.toBeNaN();
    expect(semimajorAxis).toMatchSnapshot();
  });

  // can handle drift rate of 0 when calculating semimajor axis
  it('should handle drift rate of 0 when calculating semimajor axis', () => {
    const driftRate = 0; // degrees per day
    const semimajorAxis = Earth.driftDegreesToSma(driftRate);

    expect(semimajorAxis).not.toBeNaN();
    expect(semimajorAxis).toMatchSnapshot();
  });
});
