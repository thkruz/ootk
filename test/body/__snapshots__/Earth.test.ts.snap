// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Earth should calculate Earth's diameter from satellite position 1`] = `NaN`;

exports[`Earth should calculate drift rate when given a semimajor axis 1`] = `5.858163121001669`;

exports[`Earth should calculate mean motion when given a semimajor axis 1`] = `0.0010780076124668337`;

exports[`Earth should calculate nutation angles when given an epoch 1`] = `-0.000022653923055009833`;

exports[`Earth should calculate nutation angles when given an epoch 2`] = `0.000039533533539525476`;

exports[`Earth should calculate nutation angles when given an epoch 3`] = `0.40903826060434395`;

exports[`Earth should calculate nutation angles when given an epoch 4`] = `0.4090777941378835`;

exports[`Earth should calculate nutation angles when given an epoch 5`] = `-0.000020780433407756886`;

exports[`Earth should calculate nutation angles when given an epoch 6`] = `2.342843536610454`;

exports[`Earth should calculate precession angles when given an epoch 1`] = `0.0026870317857958003`;

exports[`Earth should calculate precession angles when given an epoch 2`] = `0.002335075451748929`;

exports[`Earth should calculate precession angles when given an epoch 3`] = `0.0026872537755354814`;

exports[`Earth should calculate semimajor axis from drift rate 1`] = `42156.361072415566`;

exports[`Earth should handle drift rate of 0 when calculating semimajor axis 1`] = `42164.16961350806`;

exports[`Earth should handle epoch after 2150 when calculating nutation angles 1`] = `-0.000022653923055009833`;

exports[`Earth should handle epoch after 2150 when calculating precession angles 1`] = `0.0026870317857958003`;

exports[`Earth should handle epoch before 1900 when calculating nutation angles 1`] = `-0.000022653923055009833`;

exports[`Earth should handle epoch before 1900 when calculating precession angles 1`] = `0.0026870317857958003`;

exports[`Earth should handle negative drift rate when calculating semimajor axis 1`] = `42171.97743162094`;
